File: frontend/src/services/testCaseService.js
import axios from 'axios'

const API_BASE_URL = '/api/v1'

const apiClient = axios.create({
  baseURL: API_BASE_URL,
  headers: {
    'Content-Type': 'application/json',
  },
})

// Request interceptor
apiClient.interceptors.request.use(
  (config) => {
    const token = localStorage.getItem('token')
    if (token) {
      config.headers.Authorization = `Bearer ${token}`
    }
    return config
  },
  (error) => {
    return Promise.reject(error)
  }
)

// Response interceptor
apiClient.interceptors.response.use(
  (response) => response.data,
  (error) => {
    console.error('API Error:', error.response?.data || error.message)
    return Promise.reject(error.response?.data || error)
  }
)

/**
 * Test Case Service
 * Frontend service for test case API
 *
 * Task: T057 [US2] Implement frontend service for test case API
 */
const testCaseService = {
  /**
   * AI Generate test cases from input
   */
  generateTestCases: async (data) => {
    return apiClient.post('/test-cases/generate', data)
  },

  /**
   * Create a new test case
   */
  createTestCase: async (caseData) => {
    return apiClient.post('/test-cases', caseData)
  },

  /**
   * Get all test cases
   */
  getAllTestCases: async (params = {}) => {
    return apiClient.get('/test-cases', { params })
  },

  /**
   * Get test case by ID
   */
  getTestCaseById: async (id) => {
    return apiClient.get(`/test-cases/${id}`)
  },

  /**
   * Get test cases by status
   */
  getTestCasesByStatus: async (status) => {
    return apiClient.get(`/test-cases`, { params: { status } })
  },

  /**
   * Get test cases by type
   */
  getTestCasesByType: async (type) => {
    return apiClient.get(`/test-cases`, { params: { type } })
  },

  /**
   * Update test case
   */
  updateTestCase: async (id, caseData) => {
    return apiClient.put(`/test-cases/${id}`, caseData)
  },

  /**
   * Approve test case
   */
  approveTestCase: async (id) => {
    return apiClient.post(`/test-cases/${id}/approve`)
  },

  /**
   * Delete test case
   */
  deleteTestCase: async (id) => {
    return apiClient.delete(`/test-cases/${id}`)
  },

  /**
   * Deduplicate test cases
   */
  deduplicateTestCases: async (testCases) => {
    return apiClient.post('/test-cases/deduplicate', testCases)
  },

  /**
   * Analyze test coverage
   */
  analyzeTestCoverage: async (testCases) => {
    return apiClient.post('/test-cases/analyze-coverage', testCases)
  },
}

export default testCaseService

File: backend/src/main/java/com/synapsetest/testmanagement/service/TestCaseService.java
package com.synapsetest.testmanagement.service;

import com.synapsetest.testmanagement.dto.TestCaseRequest;
import com.synapsetest.testmanagement.dto.TestCaseResponse;
import com.synapsetest.testmanagement.exception.ResourceNotFoundException;
import com.synapsetest.testmanagement.model.TestCase;
import com.synapsetest.testmanagement.repository.TestCaseRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

/**
 * TestCase Service
 * Business logic for test case management
 *
 * Task: T048 [US2] Implement TestCaseService
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class TestCaseService {

    private final TestCaseRepository testCaseRepository;

    /**
     * Create a new test case
     */
    @Transactional
    public TestCaseResponse createTestCase(TestCaseRequest request, String username) {
        log.info("Creating test case: {} by user: {}", request.getTitle(), username);

        TestCase testCase = new TestCase();
        testCase.setTitle(request.getTitle());
        testCase.setDescription(request.getDescription());
        testCase.setSteps(request.getSteps());
        testCase.setExpectedResults(request.getExpectedResults());
        testCase.setPriority(request.getPriority() != null ? request.getPriority() : 0);
        testCase.setType(request.getType());
        testCase.setStatus(TestCase.Status.DRAFT.name());
        testCase.setTags(request.getTags());
        testCase.setRelatedRequirement(request.getRelatedRequirement());
        testCase.setCreatedBy(username);

        TestCase saved = testCaseRepository.save(testCase);

        log.info("Test case created successfully with ID: {}", saved.getId());

        return convertToResponse(saved);
    }

    /**
     * Get test case by ID
     */
    public TestCaseResponse getTestCaseById(UUID id) {
        TestCase testCase = testCaseRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("TestCase", "id", id));

        return convertToResponse(testCase);
    }

    /**
     * Get all test cases with pagination
     */
    public Page<TestCaseResponse> getAllTestCases(Pageable pageable) {
        return testCaseRepository.findAll(pageable)
                .map(this::convertToResponse);
    }

    /**
     * Get test cases by status
     */
    public List<TestCaseResponse> getTestCasesByStatus(String status) {
        return testCaseRepository.findByStatus(status)
                .stream()
                .map(this::convertToResponse)
                .collect(Collectors.toList());
    }

    /**
     * Get test cases by type
     */
    public List<TestCaseResponse> getTestCasesByType(String type) {
        return testCaseRepository.findByType(type)
                .stream()
                .map(this::convertToResponse)
                .collect(Collectors.toList());
    }

    /**
     * Update test case
     */
    @Transactional
    public TestCaseResponse updateTestCase(UUID id, TestCaseRequest request) {
        TestCase testCase = testCaseRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("TestCase", "id", id));

        testCase.setTitle(request.getTitle());
        testCase.setDescription(request.getDescription());
        testCase.setSteps(request.getSteps());
        testCase.setExpectedResults(request.getExpectedResults());
        testCase.setPriority(request.getPriority());
        testCase.setType(request.getType());
        testCase.setTags(request.getTags());
        testCase.setRelatedRequirement(request.getRelatedRequirement());

        TestCase updated = testCaseRepository.save(testCase);

        log.info("Test case updated: {}", id);

        return convertToResponse(updated);
    }

    /**
     * Approve test case
     */
    @Transactional
    public TestCaseResponse approveTestCase(UUID id) {
        TestCase testCase = testCaseRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("TestCase", "id", id));

        testCase.setStatus(TestCase.Status.APPROVED.name());
        TestCase updated = testCaseRepository.save(testCase);

        log.info("Test case approved: {}", id);

        return convertToResponse(updated);
    }

    /**
     * Delete test case
     */
    @Transactional
    public void deleteTestCase(UUID id) {
        if (!testCaseRepository.existsById(id)) {
            throw new ResourceNotFoundException("TestCase", "id", id);
        }

        testCaseRepository.deleteById(id);
        log.info("Test case deleted: {}", id);
    }

    /**
     * Convert entity to response DTO
     */
    private TestCaseResponse convertToResponse(TestCase testCase) {
        TestCaseResponse response = new TestCaseResponse();
        response.setId(testCase.getId());
        response.setTitle(testCase.getTitle());
        response.setDescription(testCase.getDescription());
        response.setSteps(testCase.getSteps());
        response.setExpectedResults(testCase.getExpectedResults());
        response.setPriority(testCase.getPriority());
        response.setType(testCase.getType());
        response.setStatus(testCase.getStatus());
        response.setTags(testCase.getTags());
        response.setRelatedRequirement(testCase.getRelatedRequirement());
        response.setCreatedAt(testCase.getCreatedAt());
        response.setUpdatedAt(testCase.getUpdatedAt());
        response.setCreatedBy(testCase.getCreatedBy());

        return response;
    }
}
